### PR TITLE
Feature: dashboard page components

### PR DESCRIPTION
**Pull request by sahsisunny**
#### Description:
This pull request addresses the implementation of individual components and utility functions required for the Dashboard page. Each component and function is designed to contribute to the overall functionality of the Dashboard.

#### File Additions:
- `src/components/Dashboard/UrlListItem.tsx` - A component that renders a single URL entry on the Dashboard page.
- `src/utils/fetchUrls.ts` - A utility function that fetches the URLs created by the logged-in user.
- `src/utils/formatDate.ts` - A utility function that formats the date in a user-friendly format.
- `src/pages/dashboard/index.tsx` -> `src/pages/app/index.tsx` - Renamed the Dashboard page to App page to reflect the new functionality.
- `public/assets/icons/internet.tsx` - Added an icon for the Dashboard page.

### Package Additions:
- `mockdate` - A package that allows for mocking the date in unit tests.

#### Testing:
| File               | Statements | Branches | Functions | Lines |
|--------------------|------------|----------|-----------|-------|
| UrlListItem.tsx    | 100        | 100      | 100       | 100   |
| fetchUrls.ts       | 100        | 100      | 100       | 100   |
| formatDate.ts      | 100        | 100      | 100       | 100   |

### CODE REVIEW

File: public/assets/icons/copy.tsx [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
+        stroke="currentColor"
+        viewBox="0 0 24 24"
+        width={width + 'px'}
+        height={height + 'px'}
+        className="text-gray-700 transition-all group-hover:text-blue-800"
+    >
+        <title>Copy</title>
+        <path d="M8 17.929H6c-1.105 0-2-.912-2-2.036V5.036C4 3.91 4.895 3 6 3h8c1.105 0 2 .911 2 2.036v1.866m-6 .17h8c1.105 0 2 .91 2 2.035v10.857C20 21.09 19.105 22 18 22h-8c-1.105 0-2-.911-2-2.036V9.107c0-1.124.895-2.036 2-2.036z"></path>
     </svg>
 );
```
Comment (ajoykumardas12) [inline public/assets/icons/copy.tsx:18]: Will it be best practice to store the TSX files in public folder?
Even though these renders svg, but they are not static files now and have more functionality as components. 
Icons Props interface is also repeated, can be a reusable interface.

**Review by ajoykumardas12 — COMMENTED** [review]
PR description should be improved (Issue number, Demo).

**Comment by sahsisunny** [comment]
> PR description should be improved (Issue number, Demo).

The issue ticket is already linked and I don't think there is any need to  add any demo videos or screenshots because In this PR, I only added some components and utility functions that are used on the Dashboard page as mentioned in the PR description

File: public/assets/icons/copy.tsx (hunk 1 above)
Comment (sahsisunny) [reply public/assets/icons/copy.tsx]: Done move all icons into components directory

File: src/utils/fetchUrls.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        });
+
+        if (response.ok) {
+            const data = (await response.json()) as { urls: UrlType[] };
+            return data.urls;
+        } else {
+            return null;
+        }
+    } catch (error) {
+        console.error('Error fetching original URL:', error);
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/utils/fetchUrls.ts:20]: Then what to do in catch block?

[the remark this answers is not in the document]
Comment (ajoykumardas12) [reply src/utils/fetchUrls.ts:20]: Is this for testing or intended for end user?
If for end user, we should show some error message in screen.

File: src/components/icons/copy.tsx [hunk 3]
```diff
@@ -0,0 +1,20 @@
+interface Props {
+    width?: number;
+    height?: number;
+}
```
Comment (ajoykumardas12) [inline src/components/icons/copy.tsx:4]: This interface can be re-used.

File: src/utils/fetchUrls.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/utils/fetchUrls.ts:20]: for testing/debugging

File: src/components/icons/copy.tsx (hunk 3 above)
Comment (sahsisunny) [reply src/components/icons/copy.tsx]: done

**Review by ajoykumardas12 — APPROVED** [review]

**Review by gauravsinhaweb — COMMENTED** [review]
What distinguishes the files in the `src/constants` path from those in `src/utils/constants`? 
 
Also,
 it is recommended not to create a folder for constants in make files. Instead, opt for a single file dedicated to constants and define all variables/constants there, irrespective of the files consuming them.

**Comment by sahsisunny** [comment]
> What distinguishes the files in the `src/constants` path from those in `src/utils/constants`?
> 
> Also, it is recommended not to create a folder for constants in make files. Instead, opt for a single file dedicated to constants and define all variables/constants there, irrespective of the files consuming them.

This change is related to this file?

**Review by gauravsinhaweb — DISMISSED** [review]
`src/constants/urls`

The APIs should be defined in the config file not in the constants.

**Comment by gauravsinhaweb** [comment]
> > What distinguishes the files in the `src/constants` path from those in `src/utils/constants`?
> > Also, it is recommended not to create a folder for constants in make files. Instead, opt for a single file dedicated to constants and define all variables/constants there, irrespective of the files consuming them.
> 
> This change is related to this file?

didn't get you, what's "this" and "this"?

**Comment by sahsisunny** [comment]
> `src/constants/urls`
> 
> The APIs should be defined in the config file not in the constants.

Show me the changes

**Comment by gauravsinhaweb** [comment]
> > `src/constants/urls`
> > The APIs should be defined in the config file not in the constants.
> 
> Show me the changes

why do you need to see it?

File: src/pages/app/index.tsx [hunk 4]
```diff
@@ -51,15 +50,15 @@ const OutputSection: React.FC<OutputSectionProps> = ({ shortUrl, handleCopyUrl }
...
+        <a
             type="button"
             className="bg-gray-200  px-2 py-4 hover:bg-gray-400"
             href={shortUrl}
             target="_blank"
             rel="noopener noreferrer"
         >
             <ShareIcon />
-        </Link>
+        </a>
```
Comment (sahsisunny) [inline src/pages/app/index.tsx:61]: I will fix this in #46 PR

Comment (vinit717) [reply src/pages/app/index.tsx:61]: Ok

**Comment by gauravsinhaweb** [comment]
I just saw you stale my comment, 
My intention is to make the codebase better, because the project is still in initial phase, later if the project gets large, it would be hard for develops optimize and make basic changes,

we should take necessary steps if the changes is required to make our codebase better 

Also, Im responsible to give valuable feedback, if i have been asked for it, I don't care if you stale it.

**Comment by sahsisunny** [comment]
> I just saw you dismissed stale my comment, fine My intention is to make the codebase better, because the project is still in initial phase, later if the project gets large, it would be hard for develops optimize and make basic changes,
> 
> we should take necessary steps if the changes is required to make our codebase better
> 
> Also, Im responsible to give valuable feedback, if i have been asked for it, it's upon you to take it or not

Yes, but you cant force me to do this in this PR. I raised the refactor PR(#50 ) I will it in refactor PR.

File: __tests__/utils/fetchUrls.test.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            credentials: 'include',
+        });
+
+        expect(result).toBeNull();
+    });
+});
+
+afterEach(() => {
+    fetchMock.resetMocks();
+});
```
Comment (vinit717) [inline __tests__/utils/fetchUrls.test.ts:55]: Nit: keep this before describe block

**Review by vinit717 — APPROVED** [review]
NIT: Try to write test for icons by giving test id to it and check whether it render or not

**Comment by gauravsinhaweb** [comment]
> > I just saw you dismissed stale my comment, fine My intention is to make the codebase better, because the project is still in initial phase, later if the project gets large, it would be hard for develops optimize and make basic changes,
> > we should take necessary steps if the changes is required to make our codebase better
> > Also, Im responsible to give valuable feedback, if i have been asked for it, it's upon you to take it or not
> 
> Yes, but you cant force me to do this in this PR. I raised the refactor PR(#50 ) I will it in refactor PR.

but you said useless comment so I had to reply, anyway

**Comment by sahsisunny** [comment]
> > > I just saw you dismissed stale my comment, fine My intention is to make the codebase better, because the project is still in initial phase, later if the project gets large, it would be hard for develops optimize and make basic changes,
> > > we should take necessary steps if the changes is required to make our codebase better
> > > Also, Im responsible to give valuable feedback, if i have been asked for it, it's upon you to take it or not
> > 
> > 
> > Yes, but you cant force me to do this in this PR. I raised the refactor PR(#50 ) I will it in refactor PR.
> 
> why I will force you XD, you said "useless comment" so I had to reply, anyway

You comment as a **Request Changes** which means this change is necessary right?

**Comment by gauravsinhaweb** [comment]
> > > > I just saw you dismissed stale my comment, fine My intention is to make the codebase better, because the project is still in initial phase, later if the project gets large, it would be hard for develops optimize and make basic changes,
> > > > we should take necessary steps if the changes is required to make our codebase better
> > > > Also, Im responsible to give valuable feedback, if i have been asked for it, it's upon you to take it or not
> > > 
> > > 
> > > Yes, but you cant force me to do this in this PR. I raised the refactor PR(#50 ) I will it in refactor PR.
> > 
> > why I will force you XD, you said "useless comment" so I had to reply, anyway
> 
> You comment as a **Request Changes** which means this change is necessary right?

Correct

**Pull request merged**